Modules/posixmodule.c

```diff
@@ -13546,46 +13550,34 @@ struct constdef {
 };
 
 static int
-conv_confname(PyObject *arg, int *valuep, struct constdef *table,
-              size_t tablesize)
+conv_confname(PyObject *module, PyObject *arg, int *valuep, const char *tablename)
 {
-    if (PyLong_Check(arg)) {
-        int value = PyLong_AsInt(arg);
-        if (value == -1 && PyErr_Occurred())
-            return 0;
-        *valuep = value;
-        return 1;
-    }
-    else {
-        /* look up the value in the table using a binary search */
-        size_t lo = 0;
-        size_t mid;
-        size_t hi = tablesize;
-        int cmp;
-        const char *confname;
-        if (!PyUnicode_Check(arg)) {
-            PyErr_SetString(PyExc_TypeError,
-                "configuration names must be strings or integers");
+    if (PyUnicode_Check(arg)) {
+        PyObject *table = PyObject_GetAttrString(module, tablename);
+        if (table == NULL) {
             return 0;
         }
-        confname = PyUnicode_AsUTF8(arg);
-        if (confname == NULL)
+
+        arg = PyObject_GetItem(table, arg);
+        Py_DECREF(table);
+        if (arg == NULL) {
+            PyErr_SetString(
+                PyExc_ValueError, "unrecognized configuration name");
             return 0;
-        while (lo < hi) {
-            mid = (lo + hi) / 2;
-            cmp = strcmp(confname, table[mid].name);
-            if (cmp < 0)
-                hi = mid;
-            else if (cmp > 0)
-                lo = mid + 1;
-            else {
-                *valuep = table[mid].value;
-                return 1;
-            }
         }
-        PyErr_SetString(PyExc_ValueError, "unrecognized configuration name");
-        return 0;
+    } else {
+        Py_INCREF(arg);  // Match the Py_DECREF below.
     }
+
+    if (PyLong_Check(arg)) {
+        *valuep = PyLong_AsInt(arg);
+    } else {
+        PyErr_SetString(PyExc_TypeError,
+            "configuration names must be strings or integers");
+    }
+
+    Py_DECREF(arg);
+    return !PyErr_Occurred();
```

Comment:
Avoid the call to `PyErr_Occurred`; instead, return NULL directly in error paths.

Reply:
OK, I've changed it to only call `PyErr_Occurred` in the unavoidable case when `PyLong_AsInt` returns -1.